dali/pipeline/operator/builtin/external_source.h

```diff
@@ -384,15 +386,24 @@ inline void ExternalSource<CPUBackend>::CopyUserData(const SourceDataType<SrcBac
 template<>
 template<typename SrcBackend, template<typename> class SourceDataType>
 inline void ExternalSource<GPUBackend>::CopyUserData(const SourceDataType<SrcBackend> &batch,
-                                                 cudaStream_t stream, bool sync) {
+                                                     cudaStream_t stream, bool sync, bool use_copy_kernel) {
   std::list<uptr_cuda_event_type> copy_to_storage_event;
   std::list<uptr_tl_type> tl_elm;
   {
     std::lock_guard<std::mutex> busy_lock(busy_m_);
     tl_elm = tl_data_.GetEmpty();
     copy_to_storage_event = copy_to_storage_events_.GetEmpty();
   }
-  tl_elm.front()->Copy(batch, stream);
+
+  constexpr bool IsSrcTensorListGPU   = std::is_same<SourceDataType<SrcBackend>, TensorList<CPUBackend>>::value;
+  constexpr bool IsSrcTensorVectorGPU = std::is_same<SourceDataType<SrcBackend>, TensorVector<CPUBackend>>::value;
+  if ((IsSrcTensorListGPU || IsSrcTensorVectorGPU) && use_copy_kernel && batch.is_pinned()) {
+    std::cout << "Using copy kernel\n";
```

Comment:
alternative: fail if the user requested use of a copy kernel but the input is non pinned host memory